### ISSUE TITLE
etcd/etcdEvents Cluster Creation Errors

### ISSUE THREAD

**Issue by mikeln** [samsung-cnct/k2#673]
EtcdEvents cluster has periodic problems when starting a new cluster.   
Also, it appears that, if the EtcdEvents cluster was failing to be established...it would never resolve the issue with time (or another up.sh).   You have to tear down and restart the cluster.

cluster error (the ec2 machines were shown as ok in the aws console):
```
mikeln-bender:.kraken Mikel_Nelson$ ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -F /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.kraken/mikeln-v17/ssh_config etcdEvents-1 etcdctl --ca-file=/etc/etcd/ssl/client-ca.pem --cert-file=/etc/etcd/ssl/client.pem --key-file=/etc/etcd/ssl/client-key.pem --endpoints=https://127.0.0.1:2381 cluster-health
Warning: Permanently added 'ec2-54-201-86-117.us-west-2.compute.amazonaws.com,54.201.86.117' (ECDSA) to the list of known hosts.
cluster may be unhealthy: failed to list members
Error:  client: etcd cluster is unavailable or misconfigured
error #0: client: endpoint https://127.0.0.1:2381 exceeded header timeout
```
vs good cluster:
```
mikeln-bender:.kraken Mikel_Nelson$ ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -F /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.kraken/mikeln-v17/ssh_config etcdEvents-1 etcdctl --ca-file=/etc/etcd/ssl/client-ca.pem --cert-file=/etc/etcd/ssl/client.pem --key-file=/etc/etcd/ssl/client-key.pem --endpoints=https://127.0.0.1:2381 cluster-health
Warning: Permanently added 'ec2-54-202-34-222.us-west-2.compute.amazonaws.com,54.202.34.222' (ECDSA) to the list of known hosts.
member 41a3e112a65a837c is healthy: got healthy result from https://10.0.78.85:2381
member a2c4d959cd59f4f6 is healthy: got healthy result from https://10.0.63.5:2381
member b7f63ca0829db36f is healthy: got healthy result from https://10.0.234.11:2381
member cd552ef50bda685e is healthy: got healthy result from https://10.0.84.207:2381
member dddc44a21e23857e is healthy: got healthy result from https://10.0.201.175:2381
cluster is healthy
```

Also, maybe related.  The `etcdctl` health checks are only run off the first node in the cluster (e.g. etcd-1 and etcdEvents-1).  This might contribute to this issue, if node 1 is having issues, and not restarted automatically. (No evidence of this yet, but a thought).

**Comment by mikeln** [comment]
more data.   Had the inverse bug...where etcdEvents was up, but etcd was down similar to above.

Logged onto the ec2 instances.  2-5 were ok.  However 1 had experienced a `kraken-etcd.service` failure several times, but had recovered and was running.  But it was continually in a `raft` vote.  excerpt:
```
ul 26 21:48:55 ip-10-0-42-160.us-west-2.compute.internal rkt[1221]: 2017-07-26 21:48:55.044129 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[2753face0de77a41]=
Jul 26 21:48:55 ip-10-0-42-160.us-west-2.compute.internal rkt[1221]: 2017-07-26 21:48:55.044590 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[2753face0de77a41]=
Jul 26 21:48:55 ip-10-0-42-160.us-west-2.compute.internal rkt[1221]: 2017-07-26 21:48:55.100452 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[8df4dd1ed83bb9a2]=
Jul 26 21:48:55 ip-10-0-42-160.us-west-2.compute.internal rkt[1221]: 2017-07-26 21:48:55.100988 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[8df4dd1ed83bb9a2]=
Jul 26 21:48:55 ip-10-0-42-160.us-west-2.compute.internal rkt[1221]: 2017-07-26 21:48:55.101054 I | raft: 909f02d7e0c7d0df [term: 11] received a MsgVote message with higher term fr
Jul 26 21:48:55 ip-10-0-42-160.us-west-2.compute.internal rkt[1221]: 2017-07-26 21:48:55.101069 I | raft: 909f02d7e0c7d0df became follower at term 12
Jul 26 21:48:55 ip-10-0-42-160.us-west-2.compute.internal rkt[1221]: 2017-07-26 21:48:55.101079 I | raft: 909f02d7e0c7d0df [logterm: 1, index: 5, vote: 0] cast MsgVote for 37abf47c
Jul 26 21:48:55 ip-10-0-42-160.us-west-2.compute.internal rkt[1221]: 2017-07-26 21:48:55.121532 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[55b48c1e43d8e46f]=
Jul 26 21:48:55 ip-10-0-42-160.us-west-2.compute.internal rkt[1221]: 2017-07-26 21:48:55.157799 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[55b48c1e43d8e46f]=
Jul 26 21:48:55 ip-10-0-42-160.us-west-2.compute.internal rkt[1221]: 2017-07-26 21:48:55.159521 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[2753face0de77a41]=
Jul 26 21:48:55 ip-10-0-42-160.us-west-2.compute.internal rkt[1221]: 2017-07-26 21:48:55.159787 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[2753face0de77a41]=
Jul 26 21:48:55 ip-10-0-42-160.us-west-2.compute.internal rkt[1221]: 2017-07-26 21:48:55.215846 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[8df4dd1ed83bb9a2]=
```

**Comment by mikeln** [comment]
How can/do we restart a single bad etcd machine?

what I just saw.. node 1 was spinning in a raft election ..but not really talking to the other 4 nodes correctly.
in that case…the other 4 nodes should still be usable..correct?…  so the etcd cluster is essentially ok…

I guess what I’m saying… we only query etcd-1 and etcdEvents-1 to get cluster health. (via ssh).  In this case etcd-1 was hosed, but I suspect 2-5 were ok and probably would have reported “cluster healthy” (though I didn’t check).


patthec:
a. we have no mechanism for recovering from that situation 
or I think david is working on writing up some manual documentation for how to resolve it
b. do we want to continue if only nodes 2-5 spin up correctly?

mikeln:
that is the question…

so we consider this as healthy (as a counter argument)…note node-4 is down:
```
mikeln-bender:.kraken Mikel_Nelson$ ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -F /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.kraken/mikeln-v17/ssh_config etcd-1 etcdctl --ca-file=/etc/etcd/ssl/client-ca.pem --cert-file=/etc/etcd/ssl/client.pem --key-file=/etc/etcd/ssl/client-key.pem --endpoints=https://127.0.0.1:2379,https://127.0.0.1:4001 cluster-health
Warning: Permanently added 'ec2-54-218-254-19.us-west-2.compute.amazonaws.com,54.218.254.19' (ECDSA) to the list of known hosts.
member 272603bce41db13c is healthy: got healthy result from https://10.0.103.43:2379
member 3db13175ab1f84f8 is healthy: got healthy result from https://10.0.212.209:2379
member 99554b54923b4718 is healthy: got healthy result from https://10.0.114.1:2379
member b173ae8654361fa6 is unreachable: no available published client urls
member d55806c86029c413 is healthy: got healthy result from https://10.0.149.146:2379
cluster is healthy
```

(but at least it has recovered… now it is up).   but my point, if etcdctl returns “healthy” then our setup proceeds.

patthec:
so based on that, we should be checking all five nodes and checking for a single valid cluster that may be missing nodes
or, checking all nodes and making sure all are healthy

mikeln:
I’m going the long way to maybe justify checking cluster health from more than node 1
yes… (the first would be better for our dev/ci times)… though the 2nd is more rigorous.

**Comment by mikeln** [comment]
To answer the speculation above.  etcdEvents cluster failed.  The status from etcdEVents-1 was bad, The status from etcdEvents-2 however was ok:  (note node 2 is first below, then node 1):
```
mikeln-bender:.kraken Mikel_Nelson$ ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -F /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.kraken/mikeln-v17/ssh_config etcdEvents-2 etcdctl --ca-file=/etc/etcd/ssl/client-ca.pem --cert-file=/etc/etcd/ssl/client.pem --key-file=/etc/etcd/ssl/client-key.pem --endpoints=https://127.0.0.1:2381 cluster-health
Warning: Permanently added 'ec2-54-244-36-229.us-west-2.compute.amazonaws.com,54.244.36.229' (ECDSA) to the list of known hosts.
member 306f013adadf31a3 is healthy: got healthy result from https://10.0.250.81:2381
member 33fdcf2c4e89fb72 is unreachable: no available published client urls
member b5a88ea4bb6c0b43 is unreachable: no available published client urls
member bb510f81801c8aa3 is healthy: got healthy result from https://10.0.86.120:2381
member fdc02f7d7dcc0f11 is healthy: got healthy result from https://10.0.10.52:2381
cluster is healthy
mikeln-bender:.kraken Mikel_Nelson$ ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -F /Users/Mikel_Nelson/dev/cloud/k2new/myconfig/.kraken/mikeln-v17/ssh_config etcdEvents-1 etcdctl --ca-file=/etc/etcd/ssl/client-ca.pem --cert-file=/etc/etcd/ssl/client.pem --key-file=/etc/etcd/ssl/client-key.pem --endpoints=https://127.0.0.1:2381 cluster-health
Warning: Permanently added 'ec2-54-212-208-182.us-west-2.compute.amazonaws.com,54.212.208.182' (ECDSA) to the list of known hosts.
cluster may be unhealthy: failed to list members
Error:  client: etcd cluster is unavailable or misconfigured
error #0: dial tcp 127.0.0.1:2381: getsockopt: connection refused
```

**Comment by mikeln** [comment]
hmm from the etcd3.1 to 3.2 doc: https://github.com/coreos/etcd/blob/master/Documentation/upgrades/upgrade_3_2.md

almost looks like maybe something related..?
```
3.2 introduces two breaking changes.

Previously, clientv3.Lease.TimeToLive API returned lease.ErrLeaseNotFound on non-existent lease ID. 3.2 instead returns TTL=-1 in its response and no error (see #7305).
```

**Comment by mikeln** [comment]
I've had a number of etcdEvents failure on node 5.   I thought it was a definite thing, but now I'm getting a different node fail.

**Comment by mikeln** [comment]
isolated example...
```
ul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.934053 I | etcdmain: etcd Version: 3.2.4
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.935094 I | etcdmain: Git SHA: c31bec0
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.935350 I | etcdmain: Go Version: go1.8.3
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.935570 I | etcdmain: Go OS/Arch: linux/amd64
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.935784 I | etcdmain: setting maximum number of CPUs to 2, total number of available CPUs is 2
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.936047 I | embed: peerTLS: cert = /etc/etcd/ssl/peer.pem, key = /etc/etcd/ssl/peer-key.pem, ca = , trusted-ca = /etc/etcd/ssl/peer-ca.pem, client-cert-auth = true
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.937025 I | embed: listening for peers on https://10.0.77.95:2382
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.937355 I | embed: listening for client requests on 10.0.77.95:2381
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.937643 I | embed: listening for client requests on 127.0.0.1:2381
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.951618 N | embed: got bootstrap from DNS for etcd-server at 0=https://i-0a526d3ed3822207c.etcdevents.mikeln-v17.internal:2382
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.951634 N | embed: got bootstrap from DNS for etcd-server at 1=https://i-0095dd1934a4936eb.etcdevents.mikeln-v17.internal:2382
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.951639 N | embed: got bootstrap from DNS for etcd-server at 2=https://i-00b3930f87d13c1be.etcdevents.mikeln-v17.internal:2382
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.951642 N | embed: got bootstrap from DNS for etcd-server at 3=https://i-053afbfbf8781ae00.etcdevents.mikeln-v17.internal:2382
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal rkt[1066]: 2017-07-28 02:18:33.951714 C | etcdmain: error setting up initial cluster: cannot find local etcd member "etcdetcdEvents10.0.77.95" in SRV records
Jul 28 02:18:33 ip-10-0-77-95.us-west-2.compute.internal systemd[1]: kraken-etcd.service: Main process exited, code=exited, status=1/FAILURE
```

**Comment by alejandroEsc** [comment]
I keep seeing what you are seeing with regard to node 5 failing. In addition it seems like it could not contact one of the nodes and keeps looping on that. To solve the issue I tried the following

- delete all etcd events (I found out that apiserver is started with A record pointing to these nodes so safe to rebuild)
- ran up.sh again

Again I had node 5 fail

So the next thing I tried was:
- stop the etcd process using `sudo kill -9 <pid>` 
- restart using systemctl
This time node 5 was able to contact what seems to be node 1 and it worked, got healthy state again.

**Comment by mikeln** [comment]
ok, I'm seeing a different error on most current runs.  TLS handshake error
```
Checking etcd cluster health ----------------------
Warning: Permanently added 'ec2-54-214-173-58.us-west-2.compute.amazonaws.com,54.214.173.58' (ECDSA) to the list of known hosts.
member 9023036a6369b19 is healthy: got healthy result from https://10.0.97.218:2379
member b0585f3cbca2adc is healthy: got healthy result from https://10.0.226.195:2379
member 1bbf2d80d8f36060 is unreachable: no available published client urls
member 80350c400ecc55b7 is healthy: got healthy result from https://10.0.171.45:2379
member f457692732094193 is healthy: got healthy result from https://10.0.107.58:2379

GOOD NODE:
0.0.1:4001 member list
9023036a6369b19: name=etcdetcd10.0.97.218 peerURLs=https://i-0a58743f16de1e11e.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.97.218:2379,https://10.0.97.218:4001 isLeader=false
b0585f3cbca2adc: name=etcdetcd10.0.226.195 peerURLs=https://i-00a070209a72b8d89.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.226.195:2379,https://10.0.226.195:4001 isLeader=true
1bbf2d80d8f36060: name=0 peerURLs=https://i-0d2a676a56afbe0f7.etcd.mikeln-v17.internal:2380 clientURLs= isLeader=false
80350c400ecc55b7: name=etcdetcd10.0.171.45 peerURLs=https://i-0c9ddbd77668ae076.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.171.45:2379,https://10.0.171.45:4001 isLeader=false
f457692732094193: name=etcdetcd10.0.107.58 peerURLs=https://i-02cf72f495d4dd77b.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.107.58:2379,https://10.0.107.58:4001 isLeader=false

BAD NODE:
core@ip-10-0-56-61 ~ $ etcdctl --ca-file=/etc/etcd/ssl/client-ca.pem --cert-file=/etc/etcd/ssl/client.pem --key-file=/etc/etcd/ssl/client-key.pem --endpoints=https://127.0.0.1:2379,https://127.0.0.1:4001 member list
Error:  net/http: TLS handshake timeout
```

**Comment by mikeln** [comment]
A VERY old issue, but describes what we are seeing here: https://github.com/coreos/etcd/issues/2539

**Comment by mikeln** [comment]
CAPTURING Old Conversations...on discovering how things work:

there is an arg --initial-cluster  <list of all advertised urls for nodes>   we aren’t setting that.
everything is looks correct though.. essentially all other args just specify the local public URL for the current etcdnode.
--initial-cluster is the only one I can find that lets you specify your whole cluster.

 ```
 CGroup: /system.slice/kraken-etcd.service
           └─1024 /usr/local/bin/etcd --data-dir /ephemeral/etcd --name etcdetcdEvents --peer-client-cert-auth --peer-trusted-ca-file=/etc/etcd/ssl/peer-ca.pem --peer-cert-file=/etc/etcd/ssl/peer.pem --peer-key-file=/etc/etcd/ssl/peer-key.pem --trusted-ca-file=/etc/etcd/ssl/server-ca.pem --cert-file=/etc/etcd/ssl/server.pem --key-file=/etc/etcd/ssl/server-key.pem --client-cert-auth --discovery-srv etcdEvents.mikeln-v17.internal --initial-advertise-peer-urls https://10.0.2.229:2382 --initial-cluster-token animism-training-chastity --initial-cluster-state new --advertise-client-urls https://10.0.2.229:2381 --listen-client-urls https://127.0.0.1:2381,https://10.0.2.229:2381 --listen-peer-urls https://10.0.2.229:2382
```

this is my etcdEvents-5 node.
(and I’ve checked the others.. and our code)..   we don’t attempt to specify an initial cluster.
I’m looking at this page: https://github.com/coreos/etcd/blob/master/Documentation/op-guide/clustering.md
but then… :doh:  we do seem to specify a discovery server…so maybe never mind

patthec [5:58 PM]  Huh.  We could be.  We know the dns entries at ansible time.
mikeln [6:01 PM] hmm… I wonder if this discovery is broken anyway… the arg appears to be different:
we have --discovery-srv,  this manual says --discovery…
also.. I don’t quite understand how this discovery server even works.. I “think” the url needs to bo accessible..but I don’t see a server start for it ..  I suspect I’m missing something.
the discovery-srv url (for me) is `etcdEvents.mikeln-v17.internal` 
ahh ok:
```
–discovery
Discovery URL used to bootstrap the cluster.
default: none
env variable: ETCD_DISCOVERY
–discovery-srv
DNS srv domain used to bootstrap the cluster.
default: none
env variable: ETCD_DISCOVERY_SRV
```

so we use DNS for discovery:
```
core@ip-10-0-91-233 ~ $ nslookup etcdEvents.mikeln-v17.internal
Server:        10.0.0.2
Address:    10.0.0.2#53

Non-authoritative answer:
Name:    etcdEvents.mikeln-v17.internal
Address: 10.0.94.146
Name:    etcdEvents.mikeln-v17.internal
Address: 10.0.204.39
Name:    etcdEvents.mikeln-v17.internal
Address: 10.0.233.77
Name:    etcdEvents.mikeln-v17.internal
Address: 10.0.2.229
Name:    etcdEvents.mikeln-v17.internal
Address: 10.0.91.233
```

patthec [6:16 PM] Yes.  We use the SRV records in route53.
mikeln [6:17 PM]  ok this looks fine… I found the section in the etcd manual for DNS with IP bootstrap
patthec [6:17 PM]  If the dns records come up slow that could cause problems.
mikeln [6:18 PM] also… the --name are we supply is the same for each node… they show it as different.
could be a problem also.
this could be it:  
```
–name
Human-readable name for this member.
default: "default"
env variable: ETCD_NAME
This value is referenced as this node's own entries listed in the --initial-cluster flag (e.g., default=http://localhost:2380). This needs to match the key used in the flag if using static bootstrapping. When using discovery, each member must have a unique name. Hostname or machine-id can be a good choice.
```

it is the same e.g.: `--name etcdetcdEvents`
should be `--name etcdEvents-5`
or similar
so that could be an issue, and slow DNS would cause issues also..orphans.
let me try fixing that name first…and see

but, if we know the IPs  we “could” statically construct the cluster
instead of discovery… more of a PITA but I would hope more reliable.

patthec [6:23 PM] Hah!  Thats a bug i maybe introduced when i was taking out the auto scaling based on aws lambda functions. 
There is an issue for handling static etcd ips.

mikeln [6:47 PM]  well, this first cut didn’t seem to help

[6:48] 
I’m waist deep in it now…
hmm here’s a similar problem … response was the port wasn’t open

ah heck, this is the problem: etcdmain: error setting up initial cluster: cannot find local etcd member “etcdetcdEvents10.0.77.95” in SRV records.  So it looks like the etcd --name HAS to match its own Route 53 SRV record.  I hope we have that in the Teraform  info ( the SRV values are like: i-0a526d3ed3822207c.etcdEvents.mikeln-v17.internal

ugg even worse… by this we need to create SRV entries like:

patthec [7:33 PM] The srv records are built in tf.  So we can pasd them wherever they need to go.
mikeln [7:33 PM] `_etcd-server-ssl._tcp.etcdEvents.mikeln-v17.internal`  (if we are using ssl/tls…) or `_etcd-server._tcp.etcdEvents.mikeln-v17.internal` for non ssl/tls…
patthec [7:34 PM] Yep.  I believe we are.
ok those SRV records are there…

this is the error that I just caught: `etcdmain: error setting up initial cluster: cannot find local etcd member "etcdetcdEvents10.0.77.95" in SRV records`
that is how I changed the `--name` value..  I had to use the IP since it was the only availble unique data.

is there a terraform substitute var like ${DEFAULT_IPV4}   for the id value:
```
"resources": {
                "aws_instance.etcdEvents_1": {
                    "type": "aws_instance",
                    "depends_on": [
                        "aws_security_group.vpc_etcd_etcdEvents_secgroup",
                        "coreosbox_ami.etcdEvents_ami"
                    ],
                    "primary": {
                        "id": "i-0095dd1934a4936eb",
```
so that is wierd… that error is only on one node… the others seem to know what they are doing.

I’m not sure how this actually works correctly: `embed: got bootstrap from DNS for etcd-server at etcdetcdEvents10.0.232.209=https://i-053afbfbf8781ae00.etcdevents.mikeln-v17.internal:2382`
those are the working ones…
but looking at route53 I see no way it can find that.
patthec [7:51 PM]  We do create those records, i thought.
mikeln [7:52 PM] not for `etcdetcdEvents10.0.232.209` the etcd --name
patthec [7:53 PM] Ahh, no.  We do build thr right aide of the equals sign.
mikeln [7:54 PM] yes… that is in the SRV record
ooo wait.. this is promising.. .like  “exceeding our usage limit”: 
sorry… here’s the relavent part:
```
etcdmain: error setting up initial cluster: srv: too many errors querying DNS SRV records ("error querying DNS SRV records for _etcd-server-ssl lookup _etcd-server-ssl._tcp.etcd.mikeln-v17.internal on 10.0.0.2:53: no such host", "error querying DNS SRV records for _etcd-server lookup _etcd-server._tcp.etcd.mikeln-v17.internal on 10.0.0.2:53: no such host")
```

so DNS wasn’t up
ahh, but it kept restarting until is worked correctly…so this is ok..

hmm so wierd.  the issue is that one (or more, but usually just one) starts itself on a different cluster ID  and then tries to be the leader, but all the others are on a different cluster ID.  both sides see each other’s requests…but ignore them because they are different cluster IDs.
I haven’t found “WHY” the odd one decides to use a different cluster ID

 ```
etcdserver: initial cluster = 0=https://i-0d49517dae78b1856.etcd.mikeln-v17.internal:2380,1=https://i-0d78caf3191bb25aa.etcd.mikeln-v17.internal:2380,2=https://i-0e5b910e8032227ce.etcd.mikeln-v17.internal:2380,3=https://i-06905cc634c0b477e.etcd.mikeln-v17.internal:2380,etcdetcd10.0.89.112=https://i-06943080cfd224ed6.etcd.mikeln-v17.internal:2380
etcdserver: initial cluster = 0=https://i-0e5b910e8032227ce.etcd.mikeln-v17.internal:2380,1=https://i-06905cc634c0b477e.etcd.mikeln-v17.internal:2380,2=https://i-06943080cfd224ed6.etcd.mikeln-v17.internal:2380,3=https://i-0d49517dae78b1856.etcd.mikeln-v17.internal:2380,etcdetcd10.0.111.80=https://i-0d78caf3191bb25aa.etcd.mikeln-v17.internal:2380
```

that was just for me… those are ones that are working…

the point above… the one with the “odd name” is the local node (it goes by that --name value).
just an observation… probably not really useful info other understranding how it works.

fyi: the change I’m running: `etcd.units.etcd.part.jinja2`
```
    --name etcd{{cluster_node_tuple.1.name}}${DEFAULT_IPV4} \
     --debug \
```

if you want to try… plus I’m running 3.2.4 etcd…

mikeln
[8:58 PM] 
```
etcd-1
core@ip-10-0-69-157 ~ $ etcdctl --ca-file=/etc/etcd/ssl/client-ca.pem --cert-file=/etc/etcd/ssl/client.pem --key-file=/etc/etcd/ssl/client-key.pem --endpoints=https://127.0.0.1:2379,https://127.0.0.1:4001 member list
111bc18ac00c2340: name=etcdetcd10.0.97.218 peerURLs=https://i-032d4ea2d5910d4cc.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.97.218:2379,https://10.0.97.218:4001 isLeader=false
2eeb828ab5cacdb1: name=etcdetcd10.0.69.157 peerURLs=https://i-06745849b8280d3b6.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.69.157:2379,https://10.0.69.157:4001 isLeader=false
6a4b1cc0165b744d: name=0 peerURLs=https://i-05c7739cca2ab98e6.etcd.mikeln-v17.internal:2380 clientURLs= isLeader=false
Add Comment Click to expand inline 13 lines
```
note “split brain” essentially

patthec [9:00 PM] 
right, I've seen this a bunch when clusters are coming up or down.  depending on which API server you hit you'll see like 8 nodes, then you hit another API server you see the rest of the nodes

mikeln [9:00 PM] 
decafbdd548453bf is the “real leader” but a411d1ff758f3dc7 thinks it’s a leader

I wonder if this is still the name thing…
e.g.
 ```
a411d1ff758f3dc7: name=etcdetcd10.0.11.137 peerURLs=https://i-0b0928ab17f431464.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.11.137:2379,https://10.0.11.137:4001 isLeader=true

a411d1ff758f3dc7: name=0 peerURLs=https://i-0b0928ab17f431464.etcd.mikeln-v17.internal:2380 clientURLs= isLeader=false
```

same etcd ID… but different names

patthec [9:03 PM]  name=0?

mikeln [9:04 PM]  yes.. the “discovered” nodes are: 0, 1, 2, 3, 4  … the local ones are etcdetcd<IP>

this is closest so far: https://stackoverflow.com/questions/40585943/etcd-cluster-id-mistmatch
stackoverflow.com
etcd cluster id mistmatch
Hey I have a cluster id mismatch for some reason, i had it on 1 node then disapperead after clearing data dir few times , changing cluster token and node names, but apperead on another here is the
 
(problem is, these are “old” instances… which we don't have)

I wonder if the 2 etcd cluster interfere with each other?  (very doubtful)… but the names 0,1,2,3,4  are the same on both.. and all the issues I find talk about this only happens awhen a new cluster is trying to start when an old one is there…

patthec [9:09 PM] the startup token is supposed to prevent that kind of crosstalk
mikeln [9:10 PM] and my look here…seems to show that is true
so… maybe the early nodes start and restart…  to the point where the later nodes LOOK like a new cluster, while the current one appears old… but the whole startup should handle that I would think.   Unless we are exceeding some timeout.. 1000ms for elections?
patthec [9:11 PM] 1 second is the timeout?
easily blowing by that
image downloads for a single cluster can vary by 3 minutes
the etcd image

mikeln [9:12 PM] 
not sure this matters…just speculating:
```
etcdserver: name = etcdetcdEvents10.0.121.72
 etcdserver: data dir = /ephemeral/etcd
etcdserver: member dir = /ephemeral/etcd/member
 etcdserver: heartbeat = 100ms
 etcdserver: election = 1000ms
|etcdserver: snapshot count = 100000
 etcdserver: advertise client URLs = https://10.0.121.72:2381
 etcdserver: initial advertise peer URLs = https://10.0.121.72:2382
```
(edited)

patthec [9:14 PM]  seems promising, in the 'it might be a problem' sort of way

mikeln [9:15 PM] 
then I read this:
```
The cluster ID is computed based on the initial-cluster string; the initial-clusters don't match so the cluster ID's don't match and etcd thinks there are two separate clusters. I was able to get tunneling to work by setting the peer advertise URLs to the tunneled URLs:
```

but that can’t be correct, as the strings are different for us for each node…discovery creates that

patthec [9:19 PM] the initial-cluster string may be the thing we set in the config file, no?

mikeln [9:19 PM]  we don’t set it… because we use discovery…but I wonder…
their docs don’t show using --inital-cluster if you use discovery…

well wait though… we supply that token in our config file..  and every one would use the same ones…
(though they shouldn’t intersect…but still?)
patthec [9:25 PM]  these clusters are all in separate VPCs.  they can't see each other
mikeln [9:25 PM] exactly…
I’ll try it.. can’t hurt… to change my own token values…

**Comment by mikeln** [comment]
Ok upped the timeouts: 
```--heartbeat-interval 500 \
     --election-timeout 5000 \
```
didn't seem to help

**Comment by mikeln** [comment]
BY HAND...last resort...(haven't tried it).. remove the offending member, and re-add it:
https://coreos.com/etcd/docs/3.2.4/op-guide/runtime-configuration.html#remove-a-member

**Comment by mikeln** [comment]
Sized to m4.xlarge.  initial test, came up bad, but righted itself.

**Comment by mikeln** [comment]
FYI...looking at the etcd code: the new cluster ID is generated from a hash of all the member IDs.
`etcd/etcdserver/server.go` block starting at line 322 does the discovery for new clusters.

**Comment by mikeln** [comment]
I think I just noticed something.

in Route53 we create 2 sets of identical SRV records for each etcd cluster.... there should only be one set, and really one set of _ssl and the second of non-ssl SRV.  

This may be the issue....as discovery would find a set then the same set...hmm... looking into this.

**Comment by mikeln** [comment]
Hmm, this smell like maybe part of this... from the etcd docs: https://github.com/coreos/etcd/blob/2724c3946eb2f3def5ed38a127be982b62c81779/Documentation/dev-internal/discovery_protocol.md#checking-the-status

```If registered members are still not enough, it will wait for left members to appear.

If the number of registered members is bigger than the expected size N, it treats the first N registered members as the member list for the cluster. If the member itself is in the member list, the discovery procedure succeeds and it fetches all peers through the member list. If it is not in the member list, the discovery procedure finishes with the failure that the cluster has been full.

In etcd implementation, the member may check the cluster status even before registering itself. So it could fail quickly if the cluster has been full.```

**Comment by mikeln** [comment]
https://github.com/coreos/etcd/blob/2724c3946eb2f3def5ed38a127be982b62c81779/Documentation/dev-internal/discovery_protocol.md#specifying-the-expected-cluster-size

**Comment by mikeln** [comment]
capturing data..writeup of discovery methods for AWS (and pitfalls): https://crewjam.com/etcd-aws/

**Comment by davidewatson** [comment]
The crewjam instructions are pretty awesome. They are also fairly complex, involving quite a bit of golang code.  Some of the other guides worth reading include:

From http://engineering.monsanto.com/2015/06/12/etcd-clustering/:

> etcd requires an initial bootstrapping to form a cluster. This can be accomplished in several ways. Initially we used the etcd discovery service, but we saw strange behavior when using this with AWS Auto Scaling Groups, namely ghost IP addresses in the list the service would return. Plus, the discovery service does not handle the post-bootstrap problem of members joining and leaving the cluster. In the end, we chose the static method to reduce dependencies on external systems.

From https://github.com/xuwang/aws-terraform:

> Cluster discovery is managed with dockerage/etcd-aws-cluster image. etcd cluster is formed by self-discovery through its auto-scaling group and then an etcd initial cluster is updated automatically to s3://AWS-ACCOUNT-CLUSTER-NAME-cloudinit/etcd/initial-cluster s3 bucket. Worker nodes join the cluster by downloading the etcd initial-cluster file from the s3 bucket during their bootstrap.

**Comment by davidewatson** [comment]
The links above are for completeness while we debug the existing problems. In the end, in with another issue, I would like to consider moving to static IP address discovery. In addition to reducing a depending on DNS for this critical piece of infrastructure, I think static IPs will make node recovery and rolling upgrades easier.

**Comment by mikeln** [comment]
so hypothesizing here... 
In a static setup, it knows how many members are expected and required.
In a internal etcd discovery setup, you specify how my expected members there will be.

In a DNS SRV discovery, how does it determine the number of members to require?  I haven't found any parameter to specify that value.   I suspect it takes the number of valid SRV->A records it finds (NOT PROVEN YET).  In our case, if a SRV or A record is not created in time, the expected cluster size would be smaller.   Then, when it finds the final SRV/A record it is past the setup time (SPECULATION) and thinks that it is a new cluster...giving it a new cluster ID...which is completely not what is expected by the original cluster.   

This is all speculation, but it somewhat fits the inconsistency of the number of nodes that are successfully put in a cluster.

**Comment by mikeln** [comment]
datapoint:  m4.larges (3.2.4 etcd) (1.7.1 kubernetes)  All ec2 instances show up and good.
etcd cluster is up,  etcdEvents...none of it is up.

**Comment by alejandroEsc** [comment]
I have had similar exprience @mikeln a week ago. Not sure what is going on with etcdEvents, arent they identical for the most part, just handling different traffic?

**Comment by mikeln** [comment]
yes, exactly the same.

**Comment by mikeln** [comment]
So is there a definitive reason why we use the discovery-srv method? i.e. why it was chosen over the other 2 methods?   I'm sure @venezia @coffeepac have some background?

**Comment by venezia** [comment]
so three options: dns srv, static ip, discovery.etcd.io
 - discovery.etcd.io was ruled out because it couldn't be done disconnected from the internet
 - DNS SRV records allowed us to be flexible with IPs / allow a new machine to take over an existing one in the cluster if need be.  

Also at the time we had etcd running w/ k2 last year, etcd was in an ASG thus we wouldn't know the ip addresses ahead of time.

Earlier, there were a few locations that used the etcd hosts, this is less so today, so maintaining that list was problematic (there are, in general, great reasons to use DNS in life)

My feeling is still "DNS isn't that hard, put in the records, wait two minutes/until the cluster has the right answer, deploy the nodes"

**Comment by mikeln** [comment]
maybe this is new, but `--discovery` can be done with a Public service, as you said. Or a custom private service you host and run yourself, and it exists only for the length of the main cluster creation.

**Comment by davidewatson** [comment]
@venezia, I agree, and during stand up, I mentioned that I am pursuing the "wait for DNS" approach now. I still believe that static IP addresses have advantages (reliability, predictability, etc.). However, due to `existing_vpc:`, they may not be an option.

**Comment by davidewatson** [comment]
After a discussion with @coffeepac, we are going to try adding a `ExecStartPre` command to the etcd service which will wait for the SVC record to be populated.

-

**Comment by mikeln** [comment]
I hope that fixes it...but I also suspect lack of a comm channel on some of these errors, or temporary loss of connection at the wrong time.

**Comment by davidewatson** [comment]
For documentation purposes, another way our current configuration can fail is if we conflict with with systems etcd2. After killing kraken-etcd.service, etcd2.service can take it's place.

```
core@ip-10-0-8-233 ~ $ curl --cacert /etc/etcd/ssl/client-ca.pem --cert /etc/etcd/ssl/client.pem --key /etc/etcd/ssl/client-key.pem https://127.0.0.1:2379/health
curl: (35) error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
core@ip-10-0-8-233 ~ $
```

When you remove the certs, the cluster reports as healthy and `ps` confirms the running etcd is not the one started by kraken-etcd.service.

```
core@ip-10-0-8-233 ~ $ curl http://127.0.0.1:2379/health
{"health": "true"}core@ip-10-0-8-233 ~ $ ps -ef | grep etcd
etcd      7323     1  0 23:06 ?        00:00:00 /usr/bin/etcd2
core      7402  2301  0 23:07 pts/0    00:00:00 grep --colour=auto etcd
core@ip-10-0-8-233 ~ $
```

**Comment by coffeepac** [comment]
do we need to uninstall the system etcd2 then?

**Comment by davidewatson** [comment]
@coffeepac: I am unsure. It might be used by locksmith. Even then, with our default `rebootStrategy: "off"`, it is usually not needed. Which makes me wonder, did I make a mistake during PD?

Given the choice between uninstalling it and choosing different ports for the k8s etcd clusters, I'd probably uninstall it. Seems less error prone that way.

**Comment by davidewatson** [comment]
The latest recreate shows a DNS discovery problem.

First, we notice that the etcd cluster is unhealthy:

> rodin:k2 dwat$ for i in `seq 1 5` ; do ssh -F ~/.kraken/dew/ssh_config etcd-$i etcdctl --ca-file=/etc/etcd/ssl/client-ca.pem --cert-file=/etc/etcd/ssl/client.pem --key-file=/etc/etcd/ssl/client-key.pem --endpoints=https://127.0.0.1:2379 cluster-health ; done
Warning: Permanently added 'ec2-52-90-189-182.compute-1.amazonaws.com,52.90.189.182' (ECDSA) to the list of known hosts.
**cluster may be unhealthy: failed to list members**
Error:  client: etcd cluster is unavailable or misconfigured
error #0: client: endpoint https://127.0.0.1:2379 exceeded header timeout
Warning: Permanently added 'ec2-34-231-225-159.compute-1.amazonaws.com,34.231.225.159' (ECDSA) to the list of known hosts.
**member 9a4a88992484eb86 is unreachable: no available published client urls**
member e222c1030c4411e7 is healthy: got healthy result from https://10.0.223.246:2379
member eeb51edefd275fa4 is healthy: got healthy result from https://10.0.117.80:2379
member f0e5b725a964c805 is healthy: got healthy result from https://10.0.179.99:2379
member fd82313579b1e72e is healthy: got healthy result from https://10.0.50.175:2379
cluster is healthy

Looking at the logs with `journalctl -u kraken-etcd`:

> Aug 16 16:35:51 ip-10-0-79-39.ec2.internal rkt[1046]: 2017-08-16 16:35:51.840020 W | discovery: couldn't resolve host i-05a9f78e14f9800ad.etcd.dew.internal.:2380 during SRV discovery

Of course, the records exist by the time we log in to investigate.

> core@ip-10-0-89-54 ~ $ dig +noall +answer SRV _etcd-client-ssl._tcp.etcd.dew.internal
_etcd-client-ssl._tcp.etcd.dew.internal. 1 IN SRV 0 0 2379 i-0568fbd2ec1ea12f0.etcd.dew.internal.
_etcd-client-ssl._tcp.etcd.dew.internal. 1 IN SRV 0 0 2379 i-0709842d6e3c73456.etcd.dew.internal.
_etcd-client-ssl._tcp.etcd.dew.internal. 1 IN SRV 0 0 2379 i-083b983a9998d67f7.etcd.dew.internal.
_etcd-client-ssl._tcp.etcd.dew.internal. 1 IN SRV 0 0 2379 i-0a4c4419f06d4b6d7.etcd.dew.internal.
_etcd-client-ssl._tcp.etcd.dew.internal. 1 IN SRV 0 0 2379 i-0b2e41886fbbd8cb0.etcd.dew.internal.
core@ip-10-0-89-54 ~ $ dig +noall +answer i-0709842d6e3c73456.etcd.dew.internal. i-083b983a9998d67f7.etcd.dew.internal. i-0a4c4419f06d4b6d7.etcd.dew.internal. i-0b2e41886fbbd8cb0.etcd.dew.internal. i-0568fbd2ec1ea12f0.etcd.dew.internal.
i-0709842d6e3c73456.etcd.dew.internal. 60 IN A	10.0.50.175
i-083b983a9998d67f7.etcd.dew.internal. 40 IN A	10.0.223.246
i-0a4c4419f06d4b6d7.etcd.dew.internal. 60 IN A	10.0.117.80
i-0b2e41886fbbd8cb0.etcd.dew.internal. 60 IN A	10.0.89.54
i-0568fbd2ec1ea12f0.etcd.dew.internal. 60 IN A	10.0.179.99
core@ip-10-0-89-54 ~ $ 

I thought I had verified that terraform will have created the records before etcd attempts to start. And #707 attempted to wait 10 minutes with a short TTL before starting etcd, to be really sure resolution would work. I'm adding an `ExecStartPre` section to verify now.

One final note, when this occurs, it appears that the node which fails to join will never be able to join unless the etcd WAL directory is moved or removed. The logs contain:

> Aug 16 16:52:51 ip-10-0-79-39.ec2.internal rkt[1046]: 2017-08-16 16:52:51.352864 E | rafthttp: request cluster ID mismatch (got 8964333b5d7fbe25 want 4c4ff980b45ef606)
Aug 16 16:52:51 ip-10-0-79-39.ec2.internal rkt[1046]: 2017-08-16 16:52:51.359531 E | rafthttp: request sent was ignored (cluster ID mismatch: peer[7e2a8aa27261fed5]=8964333b5d7fbe25, local=4c4ff980b45ef606)

**Comment by mikeln** [comment]
definite split brain here... note how etcd1 thinks it is master, but etcd2 thinks it is master also.
with etcd v3.2.5.  (not sure if the etcd changes are in the latest today (monday 8/21):
```
Checking etcd cluster health ----------------------
Warning: Permanently added 'ec2-54-212-214-69.us-west-2.compute.amazonaws.com,54.212.214.69' (ECDSA) to the list of known hosts.
member 7eb3d87fad908002 is healthy: got healthy result from https://10.0.64.222:2379
member c822832fdaf3a7ae is healthy: got healthy result from https://10.0.241.162:2379
cluster is healthy
return 0
Warning: Permanently added 'ec2-54-203-2-103.us-west-2.compute.amazonaws.com,54.203.2.103' (ECDSA) to the list of known hosts.
member 29cf048e67c3ff63 is healthy: got healthy result from https://10.0.5.169:2379
member 351cf1099e18c133 is healthy: got healthy result from https://10.0.172.200:2379
member 398e18d768afe9e3 is healthy: got healthy result from https://10.0.102.224:2379
member 7eb3d87fad908002 is unreachable: no available published client urls
member c822832fdaf3a7ae is unreachable: no available published client urls
cluster is healthy
return 0
Warning: Permanently added 'ec2-54-245-159-241.us-west-2.compute.amazonaws.com,54.245.159.241' (ECDSA) to the list of known hosts.
member 29cf048e67c3ff63 is healthy: got healthy result from https://10.0.5.169:2379
member 351cf1099e18c133 is healthy: got healthy result from https://10.0.172.200:2379
member 398e18d768afe9e3 is healthy: got healthy result from https://10.0.102.224:2379
member 7eb3d87fad908002 is unreachable: no available published client urls
member c822832fdaf3a7ae is unreachable: no available published client urls
cluster is healthy
return 0
Warning: Permanently added 'ec2-34-212-170-165.us-west-2.compute.amazonaws.com,34.212.170.165' (ECDSA) to the list of known hosts.
member 29cf048e67c3ff63 is healthy: got healthy result from https://10.0.5.169:2379
member 351cf1099e18c133 is healthy: got healthy result from https://10.0.172.200:2379
member 398e18d768afe9e3 is healthy: got healthy result from https://10.0.102.224:2379
member 7eb3d87fad908002 is unreachable: no available published client urls
member c822832fdaf3a7ae is unreachable: no available published client urls
cluster is healthy
return 0
Warning: Permanently added 'ec2-54-186-49-89.us-west-2.compute.amazonaws.com,54.186.49.89' (ECDSA) to the list of known hosts.
member 7eb3d87fad908002 is healthy: got healthy result from https://10.0.64.222:2379
member c822832fdaf3a7ae is healthy: got healthy result from https://10.0.241.162:2379
cluster is healthy
return 0
---------------------------------------------------
Checking etcdEvents cluster health ----------------
Warning: Permanently added 'ec2-54-213-112-64.us-west-2.compute.amazonaws.com,54.213.112.64' (ECDSA) to the list of known hosts.
member 20e596e402b59c66 is healthy: got healthy result from https://10.0.110.169:2381
member 50cdcc107b46c4e9 is healthy: got healthy result from https://10.0.12.129:2381
member 87c9daba2536b9cf is healthy: got healthy result from https://10.0.71.177:2381
member c7151ee5a3a8c129 is healthy: got healthy result from https://10.0.146.76:2381
cluster is healthy
return 0
Warning: Permanently added 'ec2-34-209-245-3.us-west-2.compute.amazonaws.com,34.209.245.3' (ECDSA) to the list of known hosts.
member 20e596e402b59c66 is healthy: got healthy result from https://10.0.110.169:2381
member 50cdcc107b46c4e9 is healthy: got healthy result from https://10.0.12.129:2381
member 87c9daba2536b9cf is healthy: got healthy result from https://10.0.71.177:2381
member c7151ee5a3a8c129 is healthy: got healthy result from https://10.0.146.76:2381
cluster is healthy
return 0
Warning: Permanently added 'ec2-34-210-83-145.us-west-2.compute.amazonaws.com,34.210.83.145' (ECDSA) to the list of known hosts.
member 20e596e402b59c66 is healthy: got healthy result from https://10.0.110.169:2381
member 50cdcc107b46c4e9 is healthy: got healthy result from https://10.0.12.129:2381
member 87c9daba2536b9cf is healthy: got healthy result from https://10.0.71.177:2381
member c7151ee5a3a8c129 is healthy: got healthy result from https://10.0.146.76:2381
cluster is healthy
return 0
Warning: Permanently added 'ec2-54-202-204-66.us-west-2.compute.amazonaws.com,54.202.204.66' (ECDSA) to the list of known hosts.
member 20e596e402b59c66 is healthy: got healthy result from https://10.0.110.169:2381
member 50cdcc107b46c4e9 is healthy: got healthy result from https://10.0.12.129:2381
member 87c9daba2536b9cf is healthy: got healthy result from https://10.0.71.177:2381
member c7151ee5a3a8c129 is healthy: got healthy result from https://10.0.146.76:2381
cluster is healthy
return 0
Warning: Permanently added 'ec2-34-211-116-134.us-west-2.compute.amazonaws.com,34.211.116.134' (ECDSA) to the list of known hosts.
cluster may be unhealthy: failed to list members
Error:  client: etcd cluster is unavailable or misconfigured
error #0: client: endpoint https://127.0.0.1:2381 exceeded header timeout

return 4
---------------------------------------------------
##################################################-
Checking etcd cluster member list 1 ----------------------
Warning: Permanently added 'ec2-54-212-214-69.us-west-2.compute.amazonaws.com,54.212.214.69' (ECDSA) to the list of known hosts.
7eb3d87fad908002: name=etcdetcd10.0.64.222 peerURLs=https://i-0ddc0ee9a063e1e88.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.64.222:2379,https://10.0.64.222:4001 isLeader=true
c822832fdaf3a7ae: name=etcdetcd10.0.241.162 peerURLs=https://i-078b2769a0bc178dc.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.241.162:2379,https://10.0.241.162:4001 isLeader=false
Checking etcd cluster member list 2 ----------------------
Warning: Permanently added 'ec2-54-203-2-103.us-west-2.compute.amazonaws.com,54.203.2.103' (ECDSA) to the list of known hosts.
29cf048e67c3ff63: name=etcdetcd10.0.5.169 peerURLs=https://i-0028467834c86e590.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.5.169:2379,https://10.0.5.169:4001 isLeader=false
351cf1099e18c133: name=etcdetcd10.0.172.200 peerURLs=https://i-0b49af3af546b1a60.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.172.200:2379,https://10.0.172.200:4001 isLeader=true
398e18d768afe9e3: name=etcdetcd10.0.102.224 peerURLs=https://i-09304a489d245ce13.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.102.224:2379,https://10.0.102.224:4001 isLeader=false
7eb3d87fad908002: name=3 peerURLs=https://i-0ddc0ee9a063e1e88.etcd.mikeln-v17.internal:2380 clientURLs= isLeader=false
c822832fdaf3a7ae: name=1 peerURLs=https://i-078b2769a0bc178dc.etcd.mikeln-v17.internal:2380 clientURLs= isLeader=false
Checking etcd cluster member list 3 ----------------------
Warning: Permanently added 'ec2-54-245-159-241.us-west-2.compute.amazonaws.com,54.245.159.241' (ECDSA) to the list of known hosts.
29cf048e67c3ff63: name=etcdetcd10.0.5.169 peerURLs=https://i-0028467834c86e590.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.5.169:2379,https://10.0.5.169:4001 isLeader=false
351cf1099e18c133: name=etcdetcd10.0.172.200 peerURLs=https://i-0b49af3af546b1a60.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.172.200:2379,https://10.0.172.200:4001 isLeader=true
398e18d768afe9e3: name=etcdetcd10.0.102.224 peerURLs=https://i-09304a489d245ce13.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.102.224:2379,https://10.0.102.224:4001 isLeader=false
7eb3d87fad908002: name=3 peerURLs=https://i-0ddc0ee9a063e1e88.etcd.mikeln-v17.internal:2380 clientURLs= isLeader=false
c822832fdaf3a7ae: name=1 peerURLs=https://i-078b2769a0bc178dc.etcd.mikeln-v17.internal:2380 clientURLs= isLeader=false
Checking etcd cluster member list 4 ----------------------
Warning: Permanently added 'ec2-34-212-170-165.us-west-2.compute.amazonaws.com,34.212.170.165' (ECDSA) to the list of known hosts.
29cf048e67c3ff63: name=etcdetcd10.0.5.169 peerURLs=https://i-0028467834c86e590.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.5.169:2379,https://10.0.5.169:4001 isLeader=false
351cf1099e18c133: name=etcdetcd10.0.172.200 peerURLs=https://i-0b49af3af546b1a60.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.172.200:2379,https://10.0.172.200:4001 isLeader=true
398e18d768afe9e3: name=etcdetcd10.0.102.224 peerURLs=https://i-09304a489d245ce13.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.102.224:2379,https://10.0.102.224:4001 isLeader=false
7eb3d87fad908002: name=3 peerURLs=https://i-0ddc0ee9a063e1e88.etcd.mikeln-v17.internal:2380 clientURLs= isLeader=false
c822832fdaf3a7ae: name=1 peerURLs=https://i-078b2769a0bc178dc.etcd.mikeln-v17.internal:2380 clientURLs= isLeader=false
Checking etcd cluster member list 5 ----------------------
Warning: Permanently added 'ec2-54-186-49-89.us-west-2.compute.amazonaws.com,54.186.49.89' (ECDSA) to the list of known hosts.
7eb3d87fad908002: name=etcdetcd10.0.64.222 peerURLs=https://i-0ddc0ee9a063e1e88.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.64.222:2379,https://10.0.64.222:4001 isLeader=true
c822832fdaf3a7ae: name=etcdetcd10.0.241.162 peerURLs=https://i-078b2769a0bc178dc.etcd.mikeln-v17.internal:2380 clientURLs=https://10.0.241.162:2379,https://10.0.241.162:4001 isLeader=false
---------------------------------------------------
Checking etcdEvents cluster member list 1 ----------------
Warning: Permanently added 'ec2-54-213-112-64.us-west-2.compute.amazonaws.com,54.213.112.64' (ECDSA) to the list of known hosts.
20e596e402b59c66: name=etcdetcdEvents10.0.110.169 peerURLs=https://i-0120bb1f1a3828720.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.110.169:2381 isLeader=false
50cdcc107b46c4e9: name=etcdetcdEvents10.0.12.129 peerURLs=https://i-00a5dc2d6c3044a89.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.12.129:2381 isLeader=false
87c9daba2536b9cf: name=etcdetcdEvents10.0.71.177 peerURLs=https://i-0b4a87ee36e321eb7.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.71.177:2381 isLeader=false
c7151ee5a3a8c129: name=etcdetcdEvents10.0.146.76 peerURLs=https://i-0615c4a357ad93f29.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.146.76:2381 isLeader=true
Checking etcdEvents cluster member list 2 ----------------
Warning: Permanently added 'ec2-34-209-245-3.us-west-2.compute.amazonaws.com,34.209.245.3' (ECDSA) to the list of known hosts.
20e596e402b59c66: name=etcdetcdEvents10.0.110.169 peerURLs=https://i-0120bb1f1a3828720.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.110.169:2381 isLeader=false
50cdcc107b46c4e9: name=etcdetcdEvents10.0.12.129 peerURLs=https://i-00a5dc2d6c3044a89.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.12.129:2381 isLeader=false
87c9daba2536b9cf: name=etcdetcdEvents10.0.71.177 peerURLs=https://i-0b4a87ee36e321eb7.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.71.177:2381 isLeader=false
c7151ee5a3a8c129: name=etcdetcdEvents10.0.146.76 peerURLs=https://i-0615c4a357ad93f29.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.146.76:2381 isLeader=true
Checking etcdEvents cluster member list 3 ----------------
Warning: Permanently added 'ec2-34-210-83-145.us-west-2.compute.amazonaws.com,34.210.83.145' (ECDSA) to the list of known hosts.
20e596e402b59c66: name=etcdetcdEvents10.0.110.169 peerURLs=https://i-0120bb1f1a3828720.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.110.169:2381 isLeader=false
50cdcc107b46c4e9: name=etcdetcdEvents10.0.12.129 peerURLs=https://i-00a5dc2d6c3044a89.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.12.129:2381 isLeader=false
87c9daba2536b9cf: name=etcdetcdEvents10.0.71.177 peerURLs=https://i-0b4a87ee36e321eb7.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.71.177:2381 isLeader=false
c7151ee5a3a8c129: name=etcdetcdEvents10.0.146.76 peerURLs=https://i-0615c4a357ad93f29.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.146.76:2381 isLeader=true
Checking etcdEvents cluster member list 4 ----------------
Warning: Permanently added 'ec2-54-202-204-66.us-west-2.compute.amazonaws.com,54.202.204.66' (ECDSA) to the list of known hosts.
20e596e402b59c66: name=etcdetcdEvents10.0.110.169 peerURLs=https://i-0120bb1f1a3828720.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.110.169:2381 isLeader=false
50cdcc107b46c4e9: name=etcdetcdEvents10.0.12.129 peerURLs=https://i-00a5dc2d6c3044a89.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.12.129:2381 isLeader=false
87c9daba2536b9cf: name=etcdetcdEvents10.0.71.177 peerURLs=https://i-0b4a87ee36e321eb7.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.71.177:2381 isLeader=false
c7151ee5a3a8c129: name=etcdetcdEvents10.0.146.76 peerURLs=https://i-0615c4a357ad93f29.etcdevents.mikeln-v17.internal:2382 clientURLs=https://10.0.146.76:2381 isLeader=true
Checking etcdEvents cluster member list 5 ----------------
Warning: Permanently added 'ec2-34-211-116-134.us-west-2.compute.amazonaws.com,34.211.116.134' (ECDSA) to the list of known hosts.
Error:  net/http: TLS handshake timeout
```

**Comment by davidewatson** [comment]
While working on a fix for #702, I encountered another error case:

> Sep 07 20:00:03 ip-10-0-120-109.ec2.internal rkt[1167]: 2017-09-07 20:00:03.029689 E | embed: couldn't resolve during SRV discovery (srv: too many errors querying DNS SRV records ("error querying DNS SRV records for _etcd-server-ssl lookup _etcd-server-ssl._tcp.etcd.
Sep 07 20:00:03 ip-10-0-120-109.ec2.internal rkt[1167]: 2017-09-07 20:00:03.030093 C | etcdmain: error setting up initial cluster: srv: too many errors querying DNS SRV records ("error querying DNS SRV records for _etcd-server-ssl lookup _etcd-server-ssl._tcp.etcd.de
Sep 07 2

> Sep 07 20:02:17 ip-10-0-120-109.ec2.internal systemd[1]: kraken-etcd.service: Failed with result 'timeout'.

> Sep 07 20:02:19 ip-10-0-120-109.ec2.internal rkt[1286]: 2017-09-07 20:02:19.147169 C | etcdmain: listen tcp 10.0.120.109:2380: bind: address already in use

I am increasing`TimeoutStartSec` for the `kraken-etcd` unit.

**Comment by davidewatson** [comment]
So far the three CI runs with #805 have succeeded, however, two anomalies were noticed.

1) Only 4 nodes were reported once:

```
[aws]                     "member ccc4adb969ad51b is healthy: got healthy result from https://10.0.17.68:2381", 
[aws]                     "member 29eb40de90b8b522 is healthy: got healthy result from https://10.0.201.214:2381", 
[aws]                     "member 53fbbef78254d7cd is healthy: got healthy result from https://10.0.251.238:2381", 
[aws]                     "member c292c21e8f658e5a is healthy: got healthy result from https://10.0.108.61:2381", 
[aws]                     "cluster is healthy"
```

One member did not have any published URLs once:

```
[aws]                     "member 2a4f4ee065fac486 is healthy: got healthy result from https://10.0.6.6:2381", 
[aws]                     "member 4a629da1f7734715 is healthy: got healthy result from https://10.0.82.254:2381", 
[aws]                     "member 60e974fe557cf40d is healthy: got healthy result from https://10.0.70.200:2381", 
[aws]                     "member 9504c4558dee4dae is unreachable: no available published client urls", 
[aws]                     "member e3c61876fb025dbf is healthy: got healthy result from https://10.0.153.1:2381", 
[aws]                     "cluster is healthy"
```
